Add UI support for custom Python operations

SamCarlberg opened the pull request:
## Screenshot

![screen shot 2016-07-26 at 3 16 41 pm](https://cloud.githubusercontent.com/assets/6320992/17152028/0cefe7ce-5344-11e6-8c6e-6090d3d7b485.png)
## Issues
- The text 'editor' is just a `TextArea`, might want to make it more full-featured
- Poor layout for the `Create New` button
- Need to add something to the wiki explaining how to make sockets (after PR gets merged)

SamCarlberg committed:
Add barebones UI support for custom Python operations

SamCarlberg committed:
PMD issues

SamCarlberg committed:
Improve 'Create New' button, make 'All Operations' expanded by default

AustinShalit commented:
I think the "Create New" button needs a margin.

When searching for operations, what do you think about getting rid of the categories and just showing operations?

SamCarlberg committed:
Minor improvements

AustinShalit commented:
Also: https://github.com/TomasMikula/RichTextFX

SamCarlberg commented:
I think it's good to keep search results separate, it makes it easier to know what category the operations belong to. Unless if it turns out that people use the search bar for the majority of the time.

SamCarlberg committed:
Improve 'Add New' button and python script template

Doesn't crash when new python scripts have errors (just alerts the user).
Loading erroneous scripts from disk will be logged but not added to the operations list

SamCarlberg committed:
Use RichTextFX for python editor

SamCarlberg commented on ui/src/main/java/edu/wpi/grip/ui/CustomOperationsListController.java:
```diff
@@ -2,114 +2,67 @@
+      loader.load();
+      return loader.getController();
+    } catch (IOException e) {
+      throw new RuntimeException("Couldn't load python editor", e);
```
Ugly

SamCarlberg commented on ui/src/main/java/edu/wpi/grip/ui/python/PythonEditorController.java:
```diff
@@ -0,0 +1,242 @@
+  private void save() {
+    if (scriptFile == null) {
+      String fileName = scriptFileName();
+      if (fileName == null) {
```
Might just want to check if `PythonScriptFile.create(scriptFileName())` throws an exception, that would also check for errors in the script

SamCarlberg commented on ui/src/main/java/edu/wpi/grip/ui/python/PythonEditorController.java:
```diff
@@ -0,0 +1,242 @@
+  private void saveAs() {
+    FileChooser chooser = new FileChooser();
+    chooser.setInitialDirectory(PythonOperationUtils.DIRECTORY);
+    chooser.setTitle("Choose ");
```
Oops, need to finish the title

SamCarlberg commented on ui/src/main/java/edu/wpi/grip/ui/python/PythonEditorController.java:
```diff
@@ -0,0 +1,242 @@
+      String code = new String(bytes, Charset.defaultCharset());
+      codeArea.replaceText(code);
+    } catch (IOException e) {
+      logger.log(Level.WARNING, "Could not read file " + file, e);
```
Should notify the user

SamCarlberg commented on ui/src/main/java/edu/wpi/grip/ui/python/PythonEditorController.java:
```diff
@@ -0,0 +1,242 @@
+          codeArea.getText().getBytes(Charset.defaultCharset())
+      );
+    } catch (IOException e) {
+      logger.log(Level.WARNING, "Could not save to " + scriptFile, e);
```
Notify the user

SamCarlberg committed:
Merge branch 'master' into feat/python-ui

SamCarlberg committed:
Alert the user if a script has an error or can't be saved/loaded

AustinShalit commented:
closes #470

AustinShalit commented on core/src/main/java/edu/wpi/grip/core/operations/python/PythonScriptFile.java:
```diff
@@ -26,6 +28,32 @@
+          + "    outputs.createNumberSocketHint(\"sum\", 0.0),\n"
+          + "]\n\n\n" // two blank lines
+          + "def perform(a, b):\n"
+          + "    return a + b\n";
```
Maybe make this a file?  What do you think?

SamCarlberg replied:
That makes some sense. I'd like to keep the contents in memory though to minimize disk accesses.

JLLeitschuh replied:
Put it in a read once file.
Cache the string in memory after its loaded.
It will make it easier to edit in the future.

AustinShalit commented on ui/src/main/java/edu/wpi/grip/ui/OperationListController.java:
```diff
@@ -31,7 +33,7 @@
   protected static final String FILTER_TEXT = "filterText";
   private final StringProperty filterText = new SimpleStringProperty(this, FILTER_TEXT, "");
-  @FXML private Tab root;
+  @FXML private TitledPane root;
```
❤️

AustinShalit commented on ui/src/main/java/edu/wpi/grip/ui/OperationListController.java:
```diff
@@ -60,10 +62,8 @@ protected void initialize() {
-        // number of matches.  This lets the user quickly see which tabs have matching operations
-        // when
-        // searching.
+        // bold and show the umber of matches.
```
umber -> number

AustinShalit commented on ui/src/main/resources/edu/wpi/grip/ui/python/python-keywords.css:
```diff
@@ -0,0 +1,31 @@
+
+.paragraph-box:has-caret {
+    -fx-background-color: #f2f9fc;
+}
```
No EOL at EOF

AustinShalit commented on ui/src/main/java/edu/wpi/grip/ui/python/PythonEditorController.java:
```diff
@@ -0,0 +1,260 @@
+  }
+
+  @FXML
+  private void openWiki() {
```
Take a look at how we did it for the [AboutDialog](https://github.com/WPIRoboticsProjects/GRIP/blob/946d8bc07ee424718ed9b28067e64fd183a7cca4/ui/src/main/java/edu/wpi/grip/ui/AboutDialogController.java#L19)

SamCarlberg commented:
I think we're going to need to use the official OpenCV python libraries instead of accessing the Java bindings through python.

SamCarlberg committed:
Improve error alerts, add some tests, make wiki button work.

Also some minor fixup from PR

JLLeitschuh commented:
Are we waiting on this for the opencv library change?

SamCarlberg commented:
Haven't had the time to figure out how to deal with the OpenCV python libraries

SamCarlberg commented:
We'd have to add the OpenCV python stuff to the maven repo:

| Group | Name |
|---|---|
| `org.opencv.python` | `python-core` |
| `org.opencv.python.natives` | `natives-win32` (etc.) |

SamCarlberg committed:
Merge branch 'master'

Conflicts:
	build.gradle
	core/src/test/java/edu/wpi/grip/core/PythonTest.java
	core/src/test/java/edu/wpi/grip/core/serialization/ProjectTest.java
	ui/src/test/java/edu/wpi/grip/ui/PaletteTest.java

SamCarlberg commented:
So this is going to be interesting. We don't need to include python bindings for OpenCV (you can just use `pip install opencv-python`). But we will need a way of converting the JavaCV mats to and from numpy arrays

JLLeitschuh commented:
Write a library to emulate numpy in jython

ianfinlay-aus commented:
I like this feature: was thinking how to dynamically add OpenCV2 calls into the tool, amd import/export the custom library for sharing

SamCarlberg commented:
Wow, this is an old PR.  IIRC it never moved forward because it's really just a python wrapper around the java libraries (which are wrappers around the C++ library...), and wouldn't support the normal python libraries people would use to interface with OpenCV like numpy or opencv2
